File: product-microservice/api/ram_api.go
package api

import (
	"product/dto"
	"product/dto/filter"
	"product/mapper"
	"product/service"
	"strconv"

	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)


type RamAPI struct {
	IRamService service.IRamService
}

func NewRamAPI(ramService service.IRamService) RamAPI {
	return RamAPI{IRamService: ramService}
}

func (ramApi *RamAPI) GetAll(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }
	
	rams := ramApi.IRamService.GetAll(page, pageSize)
	c.JSON(http.StatusOK, mapper.ToRamDTOs(rams))
}

func (ramApi *RamAPI) GetNumberOfRecords(c *gin.Context) {
	numberOfRecords := ramApi.IRamService.GetNumberOfRecords()
	c.JSON(http.StatusOK, numberOfRecords)
}

func (ramApi *RamAPI) GetByID(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	ram, err := ramApi.IRamService.GetById(id)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToRamDTO(ram))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (ramApi *RamAPI) SearchByName(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	rams, err := ramApi.IRamService.SearchByName(page, pageSize, c.Query("name"))
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToRamDTOs(rams))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (ramApi *RamAPI) GetNumberOfRecordsSearch(c *gin.Context) {
	numberOfRecords := ramApi.IRamService.GetNumberOfRecordsSearch(c.Query("name"))
	c.JSON(http.StatusOK, numberOfRecords)
}

func (ramApi *RamAPI) Filter(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	var filter filter.RAMFilter
	err = c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	rams, err := ramApi.IRamService.Filter(page, pageSize, filter)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToRamDTOs(rams))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (ramApi *RamAPI) GetNumberOfRecordsFilter(c *gin.Context) {
	var filter filter.RAMFilter
	err := c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	numberOfRecords := ramApi.IRamService.GetNumberOfRecordsFilter(filter)
	c.JSON(http.StatusOK, numberOfRecords)
}

func (ramApi *RamAPI) GetManufacturers(c *gin.Context) {
	manufacturers := ramApi.IRamService.GetManufacturers()
	c.JSON(http.StatusOK, manufacturers)
}

func (ramApi *RamAPI) GetCapacities(c *gin.Context) {
	capacities := ramApi.IRamService.GetCapacities()
	c.JSON(http.StatusOK, capacities)
}

func (ramApi *RamAPI) GetMemoryTypes(c *gin.Context) {
	memoryTypes := ramApi.IRamService.GetMemoryTypes()
	c.JSON(http.StatusOK, memoryTypes)
}


func (ramApi *RamAPI) GetSpeeds(c *gin.Context) {
	speeds := ramApi.IRamService.GetSpeeds()
	c.JSON(http.StatusOK, speeds)
}

func (ramApi *RamAPI) Create(c *gin.Context) {
	var ramDTO dto.RamDTO
	err := c.BindJSON(&ramDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	ram := mapper.ToRam(ramDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	error := ramApi.IRamService.Create(ram)

	if error == nil {
		c.JSON(http.StatusOK, "RAM stored successfully.")
	} else {
		c.JSON(http.StatusBadRequest, error.Error())
	}
}

func (ramApi *RamAPI) Update(c *gin.Context) {
	var ramDTO dto.RamDTO
	err := c.BindJSON(&ramDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	error := ramApi.IRamService.Update(ramDTO)

	if error == nil {
		c.JSON(http.StatusOK, "RAM updated successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, error.Error())
	} 
}

func (ramApi *RamAPI) Delete(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	error := ramApi.IRamService.Delete(id)

	if error == nil {
		c.JSON(http.StatusOK, "RAM deleted successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, error.Error())
	}
}

File: api-gateway/api/product_microservice/ram_api.py
import requests
from flask import request, jsonify
from utils import token_utils, role
from utils.routes.product_microservice import ram_api_routes


def get_all():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    try:
        r = requests.get(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.GET_ALL
                         + "?page={}&pageSize={}".format(page, page_size))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_number_of_records():
    try:
        r = requests.get(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.GET_NUMBER_OF_RECORDS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_by_id(id):
    try:
        r = requests.get(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.GET_BY_ID + "/{}".format(id))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def search_by_name():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    name = args.get("name")
    try:
        r = requests.get(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.SEARCH_BY_NAME +
                         "?page={}&pageSize={}&name={}".format(page, page_size, name))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_number_of_records_search():
    args = request.args.to_dict()
    name = args.get("name")
    try:
        r = requests.get(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.GET_NUMBER_OF_RECORDS_SEARCH + "?name={}".format(name))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def filter():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    data = request.json
    try:
        r = requests.post(ram_api_routes.BASE + ram_api_routes.API +
                          ram_api_routes.FILTER +
                          "?page={}&pageSize={}".format(page, page_size), json=data)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_number_of_records_filter():
    data = request.json
    try:
        r = requests.post(ram_api_routes.BASE + ram_api_routes.API +
                          ram_api_routes.GET_NUMBER_OF_RECORDS_FILTER, json=data)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_manufacturers():
    try:
        r = requests.get(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.GET_MANUFACTURERS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_capacities():
    try:
        r = requests.get(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.GET_CAPACITIES)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_memory_types():
    try:
        r = requests.get(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.GET_MEMORY_TYPES)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_speeds():
    try:
        r = requests.get(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.GET_SPEEDS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def create():
    data = request.json
    headers = request.headers
    try:
        r = requests.post(ram_api_routes.BASE + ram_api_routes.API +
                          ram_api_routes.CREATE, json=data, headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def update():
    data = request.json
    headers = request.headers
    try:
        r = requests.put(ram_api_routes.BASE + ram_api_routes.API +
                         ram_api_routes.UPDATE, json=data, headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def delete(id):
    headers = request.headers
    try:
        r = requests.delete(ram_api_routes.BASE + ram_api_routes.API +
                            ram_api_routes.DELETE + "/{}".format(id), headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp
